src/uploaders/BrowserUploader.ts

```diff
@@ -0,0 +1,141 @@
+import { promises as fs } from 'fs'
+import path from 'path'
+import http from 'http'
+import readPkgUp from 'read-pkg-up'
+
+import { Logger, noopLogger } from '../Logger'
+import request from '../Request'
+import { UploadErrorCode, UploadError } from '../UploadError'
+import AddSources from '../transformers/AddSources'
+import StripProjectRoot from '../transformers/StripProjectRoot'
+
+interface UploadOpts {
+  apiKey: string,
+  sourceMap: string,
+  bundleUrl: string,
+  bundle?: string,
+  appVersion?: string,
+  overwrite?: boolean
+  projectRoot?: string
+  endpoint?: string
+  requestOpts?: http.RequestOptions
+  logger?: Logger
```

Comment:
Do these need to optional if they have default values when they're used? It seems like it's impossible for them to actually be missing but I@m not sure if this is a TS limitation

Reply:
It allows someone calling the library from TS (which is a use case) to _not_ set the optional ones when passing `opts`.

I think if we made them non-optional, this would not compile?

```ts
await uploadOne({
  apiKey: '123',
  sourceMap: 'dist/bundle.js.map',
  bundleUrl: 'http://my.url/static'
})
```

Reply:
I'm not sure, that's what I meant by "TS limitation" — logically it should work because if you omit the properties then they are added by the default values, but I don't know if they TS compiler allows that

It just seems weird that they are optional when they aren't (the function that uses the type needs those values) but it's not the end of the world

Reply:
If I make the types non optional on the `UploadOpts` interface, vscode says this:

![image](https://user-images.githubusercontent.com/609579/97431411-296c4800-1912-11eb-8685-baeaa4fda4a5.png)